File: Sources/UITesting/Activity.swift
import XCTest

extension State {

  public consuming func activity<Destination: List>(
    _ name: String,
    perform action: (consuming Self) throws -> State<Destination>
  ) rethrows -> State<Destination> {

    // State is non-copyable, so we must create a new State inside the
    // XCTContext.runActivity closure. Also on the way out the state must be
    // recreated because it can't be copied out of the closure.

    let application = self.application
    let screens = self.screens

    let destination = try XCTContext.runActivity(named: name) { _ in
      let state = State(application: application, screens: screens)
      return try action(state).screens
    }

    return State<Destination>(application: application, screens: destination)
  }
}

File: src/main/kotlin/com/marksandspencer/uitesting/Activity.kt
package com.marksandspencer.uitesting

import android.util.Log

fun <Content : Screen, Return : Screen> State<Content>.activity(
    name: String,
    action: (State<Content>) -> State<Return>,
): State<Return> {
    Log.d("UITesting", "Activity: $name — start")
    return action(this).also {
        Log.d("UITesting", "Activity: $name — end")
    }
}
